### DISCUSSION TITLE
KeyErrors using bound parameters in case statement

### DISCUSSION THREAD

**Discussion by ashiner** [sqlalchemy/sqlalchemy#8017] [Usage Questions] [answered]
I have an application where I am I create a recursive CTE and map the corresponding selectable to a table.  I use bound parameters to control the filter conditions in the where clause of the anchor query.  I am now trying to use a case statement in the where clause and get KeyErrors if the same bound parameter is used in both the condition as well as the result for the case statement.  

Reusing the same parameter multiple times appears to be ok.  The issue seems to come from using the same parameter for both the condition and the result calculation in a case statement. 

To Reproduce:  

```python
import sqlalchemy as sa
import sqlalchemy.orm
import sqlalchemy.ext.declarative
import string
import traceback

DB_PATH = 'postgresql://postgres:postgres@localhost:5432/test_db'
engine = sa.create_engine(DB_PATH, echo=False)
Session = sa.orm.sessionmaker(bind=engine)
DBSession = Session()
Base = sa.ext.declarative.declarative_base()

DBSession.execute(r"drop table if exists test_table;")
DBSession.commit()

class TestTable(Base):
    __tablename__ = 'test_table'
    name = sa.Column(sa.Text, primary_key=True)
    ix = sa.Column(sa.BigInteger, autoincrement=True)

    def __init__(self, name):
        self.name = name


TestTable.__table__.create(engine)
DBSession.bulk_save_objects([TestTable(k) for k in string.ascii_uppercase])
DBSession.commit()

# Case where CTE uses a conditional where clause with the same bound parameter used for the condition as well as
# for calculating the case statement result.
# This fails with KeyError: 'names' on the bound parameter key.
try:
    # This Fails
    s1 = sa.select(TestTable.name, TestTable.ix). \
        where(
              sa.case(
                  (sa.bindparam('names', required=False, value=None) != None,
                   TestTable.name.in_(sa.bindparam('names', required=False, value=None))),
                  else_=True)
              ). \
        cte(name="s1", recursive=True).prefix_with("NOT MATERIALIZED")
    rslt = DBSession.query(s1).params({'names': ['A', 'E']}).all()
    print(*rslt, sep='\n')
except Exception as e:
    traceback.print_exc()

# Same query where the bound parameter is used for calculating the result of the case statement but is not used for the
# condition.  This works as expected.
try:
    s1 = sa.select(TestTable.name, TestTable.ix). \
        where(
              sa.case(
                  (True,
                   TestTable.name.in_(sa.bindparam('names', required=False, value=None))),
                  else_=True)
              ). \
        cte(name="s1", recursive=True).prefix_with("NOT MATERIALIZED")

    rslt = DBSession.query(s1).params({'names': ['A', 'E']}).all()
    print(*rslt, sep='\n')
except Exception as e:
    traceback.print_exc()

# Same query where the bound parameter is used for calculating the condition in the cases statement but is not used for
# calculating the result.  This also works as expected.
try:
    s1 = sa.select(TestTable.name, TestTable.ix). \
        where(
              sa.case(
                  (sa.bindparam('names', required=False, value=None) != None,
                   True),
                  else_=True)
              ). \
        cte(name="s1", recursive=True).prefix_with("NOT MATERIALIZED")

    rslt = DBSession.query(s1).params({'names': ['A', 'E']}).all()
    print(*rslt, sep='\n')
except Exception as e:
    traceback.print_exc()
```

Versions

OS: Windows 10
Python: 3.9
SQLAlchemy: 1.4.36
Database: PostgreSQL 13.6

Thank you for your help,

-Andrew Shiner

**Comment by zzzeek** [comment]
will try momentarily but first you should not query like this:

```py
rslt = DBSession.query(s1).params({'names': ['A', 'E']}).all()
```

that's an incorrect usage of query() which in this case I guess will create a subquery, that would be unnecessarily inefficient.  the correct way to call this is:

```py
rslt = session.execute(s1, {'names': ['A', 'E']}).all()
```

**Reply by zzzeek** [reply]
OK these are CTEs, so 

```py
session.execute(select(s1), {'names': ['A', 'E']})
```

stay away from session.query() and definitely query.params().

the remaining issue seems to be that you are creating the same bound parameter name twice in two different contexts, which is not supported.  will see if this is supposed to raise a better error and if so why not.

**Reply by ashiner** [reply]
> that's an incorrect usage of query() which in this case I guess will create a subquery, that would be unnecessarily inefficient. the correct way to call this is:
> 
> ```python
> rslt = session.execute(s1, {'names': ['A', 'E']}).all()
> ```

In my real application I have mapped the selectable with the CTE to a table with code that looks something like this: 

```python
 class TestCTE(Base):
        __table__ = s1
        __mapper_args__ = {
            'primary_key': [s1.c.name, s1.c.ix]
        }
  ``` 

I then use that table as part of more complex queries involving joins and filters.    For that application is it appropriate to use a session queries? 

rslt = DBSession.query(TestCTE).params({'names': ['A', 'E']}).all()

**Reply by zzzeek** [reply]
I originally missed that it was  CTE.  it's best to use select() in the 1.4 series as this is the modern way to query, but I was mistaken that your usage form was incorrect.  "s1" led me to believe this was a select() object directly.

**Comment by zzzeek** [comment]
You can't use the same bound parameter name in an "expanding" context and a "non-expanding" context at the same time.  "expanding" refers to the use of the parameter within IN().  The next 1.4 release will raise an informative CompileError in this case.

```py
s1 = (
    select(TestTable.name, TestTable.ix)
    .where(
        case(
            (
                bindparam("names1") != None,
                TestTable.name.in_(
                    bindparam("names2", expanding=True)
                ),
            ),
            else_=True,
        )
    )
    .cte(name="s1", recursive=True)
    .prefix_with("NOT MATERIALIZED")
)
rslt = session.execute(select(s1), {"names1": ["A", "E"], "names2": ["A", "E"]}).all()
print(*rslt, sep="\n")

```

**Reply by ashiner** [reply]
I have defined a table that incapsulates a recursive cte.  In writing queries that use the CTE I sometimes want to apply filtering in the anchor query.  So far the only way I have been able to find to get a filter condition into anchor query is with bound parameters.  

Is there a recommended way to achieve the conditional logic where: 

- if names is not defined do nothing 
- if names is a string filter on TestTable.name == :names 
- if names is a list filter on TestTable.name.in_( :names )

**Reply by zzzeek** [reply]
> I have defined a table that incapsulates a recursive cte.

can you be more specific what this means?  do you mean, you mapped a class to a select() that includes a cte() ?

>  In writing queries that use the CTE I sometimes want to apply filtering in the anchor query. So far the only way I have been able to find to get a filter condition into anchor query is with bound parameters.
> 
> Is there a recommended way to achieve the conditional logic where:
> 
>     * if names is not defined do nothing
> 
>     * if names is a stringer filter on TestTable.name == :names
> 
>     * if names is a list filter on TestTable.name.in_( :names )

I'd need to see what you're doing.  usually I would not try to have a particular select() that's fixed, I'd create the select() that I want each time using a function.  However, if you want to have the select() of fixed structure and use bound parameters as you are illustrating, just use different names for the places that you are using the parameter with in_() vs. those that you are not.

**Reply by zzzeek** [reply]
most cases where people "map a class to a statement" can be written in a more ad-hoc fashion using [aliased](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#relationship-to-aliased-class); you still get instances of your class back and you can map it to ad-hoc statements pretty flexibly.

**Reply by ashiner** [reply]
> can you be more specific what this means?

An example is shown in this question:  https://github.com/sqlalchemy/sqlalchemy/discussions/6656

My database contains a fairly deep tree structure.   I use a class like TestCTE to relate between levels of the tree and then I join the leaves to other tables.  It works really well, but now I need more control over the filtering conditions in the anchor query and other conditional logic.  The real version of TestCTE is a very handy building block that is used for building up more complex queries.  

It could be that I'm mapping the selectable to a table unnecessarily.

**Reply by zzzeek** [reply]
in #6656 I see nothing special about `TestCTE` as a class that you couldn't do just using the statement directly.   I can rewrite it as:

```py
TestCTE = outer_cte
```

and it works.   

if changing just the bound parameter is not flexible enough then I would write a function that generates the statement I want for each case, rather than using a fixed statement object each time.  Or that returns one of several fixed statements.

As before, to the degree that your real world case is doing things with TestCTE that are class-specific, like getting a sub-namespace in your result or linking to or from a relationship, you can use aliased() for that, getting back instances of the same class but using different statements as needed.

**Reply by ashiner** [reply]
Thank you.  I think I see the way forward.  

This was all very helpful.  

-Andrew

**Reply by zzzeek** [reply]
it's a little weird becasue you dont have a base table here.  But I can make one up and use adapt_on_names:

```py
class TestCTECls(Base):
    __table__ = select(
        column("start_ix"),
        column("name"),
        column("ix")
    ).subquery()

    __mapper_args__ = {"primary_key": [__table__.c.name]}

    def __repr__(self):
        return str(self)

    def __str__(self):
        return f"TestTable: start_ix={self.start_ix}, name={self.name}, ix={self.ix}"

TestCTE = aliased(TestCTECls, outer_cte, adapt_on_names=True)

```

might be something better than that though.

**Comment by CaselIT** [comment, marked as answer]
@zzzeek setting `expaning=True` seems to work. Should it not?

```py
    # This Fails
    s1 = sa.select(TestTable.name, TestTable.ix). \
        where(
              sa.case(
                  (sa.bindparam('names', required=False, value=None, expanding=True) != None,
                   TestTable.name.in_(sa.bindparam('names', required=False, value=None, expanding=True))),
                  else_=True)
              ). \
        cte(name="s1", recursive=True).prefix_with("NOT MATERIALIZED")
    rslt = DBSession.query(s1).params({'names': ['A', 'E']}).all()
    print(*rslt, sep='\n')
```

This executes the following sql that seems reasonable
```sql
WITH RECURSIVE s1(name, ix) AS NOT MATERIALIZED
(SELECT test_table.name AS name, test_table.ix AS ix
FROM test_table
WHERE CASE WHEN ((%(names_1)s, %(names_2)s) IS NOT NULL) THEN test_table.name IN (%(names_1)s, %(names_2)s) ELSE %(param_1)s END)
 SELECT s1.name AS s1_name, s1.ix AS s1_ix
FROM s1
```
with params `{'param_1': True, 'names_1': 'A', 'names_2': 'E'}`

**Reply by ashiner** [reply]
That worked. 

Thanks!

**Reply by zzzeek** [reply]
oh, right, I didnt even look to see *why* you were putting this bound parameter in the case statement.

it's interesting that it "works", I find that SQL to be bizarre, but apparently this is what it means:

```sql
test=> SELECT 1 where (NULL, NULL) IS NULL;
 ?column? 
----------
        1
(1 row)
```

but it does ...so yes, you can do that.

It's not clear if there's a way for bindparam() to figure this out without passing "expanding" but unless you pass the datatype as `Tuple()` (which you probably could) it seems you would need to set "expanding" here.

**Reply by CaselIT** [reply]
I guess with the error you are planning on adding it may be clearer. It could mention that expanding may be manually set as an alternative to using different named bindparams
